api/resolvers/wallet.js

```diff
@@ -328,9 +319,32 @@ export default {
   }
 }
 
-async function createWithdrawal (parent, { invoice, maxFee }, { me, models, lnd }) {
+async function createWithdrawal (parent, { invoice, maxFee, tokens }, { me, models, lnd }) {
   await ssValidate(withdrawlSchema, { invoice, maxFee })
 
+  const invoiceHashMap = []
+  if (!me && tokens) {
+    for (const token of tokens) {
+      const [oHash, hmac] = token.split('|')
+      if (!hmac) {
+        throw new GraphQLError('hmac required', { extensions: { code: 'BAD_INPUT' } })
+      }
+      if (createInvoiceHmac(oHash) !== hmac) {
+        throw new GraphQLError('bad hmac', { extensions: { code: 'FORBIDDEN' } })
+      }
+      // xor last half of invoice hash with HMAC of first half.
+      // this makes the invoice not usable while withdrawal is in process.
+      // if withdrawal failed, we can reverse this operation.
+      const xHash = xor(
+        Buffer.from(oHash, 'hex'),
+        Buffer.concat([
+          Buffer.alloc(16), Buffer.from(createInvoiceHmac(oHash.slice(0, 32)), 'hex').slice(0, 16)
+        ])
+      ).toString('hex')
```

Comment:
`new_hash = hash[:16] || (hash[16:] ^ HMAC_k(hash[:16])[:16])`

This is kind of a hack. But is it a good hack? :thinking: 

This meets the following requirements afaict:
- anons can't use payment tokens again while withdrawal is in process
- payment tokens can be used again if withdrawal failed since XOR is reverseable using `hash[:16]`
- anons can't predict the new hash since the HMAC uses a secret key. Knowing the HMAC of the full hash does not give them any knowledge of the HMAC of the first half (need to double-check this)